### ISSUE TITLE
Resolver error with fastify v3

### ISSUE THREAD

**Issue by adrai** [SkeLLLa/fastify-oas#30]
made a quick test....
changed `fastify: '>=2.0.0 <2.12.0 || >=2.12.1 <4.0.0'`

when opening swagger ui, and clicking on any route having a `$ref` in the schema:

![image](https://user-images.githubusercontent.com/1086194/80840078-c6b52080-8bfc-11ea-88be-9ffacf0faf8d.png)

example:
![image](https://user-images.githubusercontent.com/1086194/80840351-64a8eb00-8bfd-11ea-957d-b848647a7b15.png)

**Comment by adrai** [comment]
other downstream libs needs probably also some changes: https://github.com/fastify/fastify-static/pull/127

**Comment by adrai** [comment]
The /documentaiton/json in fastify v2 looked like this:
![image](https://user-images.githubusercontent.com/1086194/81053305-25acbb00-8ec5-11ea-87ba-88ce6e4974fc.png)

and with fastify v3 (+ $ref) it looks like this:
![image](https://user-images.githubusercontent.com/1086194/81053367-3eb56c00-8ec5-11ea-9f9e-59a94d3f455b.png)

maybe @Eomm can help?

**Comment by SkeLLLa** [comment]
@adrai It didn't resolved in on route hook, it resolved after, when `app.ready()` resolved. So the route object properties was replaced by resolved schema.

Here are some links where you can find more details:
https://github.com/fastify/fastify/issues/2123 (this is an open issue)
https://github.com/fastify/fastify/issues/2104 (this one has explanation how it works in details)

**Comment by adrai** [comment]
> It didn't resolved in on route hook, it resolved after, when `app.ready()` resolved

ok, nevertheless... this is not getting resolved anymore...

> also fixing this could be copy-pasted to fastify-swagger module, cause it uses absolutely the same approach.

hehe so if fastify-oas or fastify-swagger get's fixed for v3 the other get's fixed too ;-)
so for fastify-swagger are you aware of this { $ref } "problem"? @SkeLLLa @mcollina
(have not verified this myself on fastify-swagger)

**Comment by Eomm** [comment]
I don't know exactly the problem here for oas.

In fastify-swagger all the schemas added with `addSchema` are not given to the `swagger-ui`, for this reason it is unable to resolve `$ref`s and this feature doesn't work in v2 and v3.
This will be fixed for fastify-v3 for sure

**Comment by mcollina** [comment]
I don't understand this issue unfortunately :(.
Can you open an issue in the fastify repo with what is not possible in v3/what is missing?

> This will be fixed for fastify-v3 for sure

I do not understand what the problem is and what the fix will be.

**Comment by adrai** [comment]
Added a test in my fastify fork:
2.x branch => https://github.com/adrai/fastify/commit/e86e23d76a544de9ade9cc96712b8525da00c60c (works)
master branch =>https://github.com/adrai/fastify/commit/0d535c4e746853aa32050db0b54ff5032abbec0d (fails)

**Comment by mcollina** [comment]
That's by design and part of some of the changes @Eomm made. I would recommend to open an issue on the fastify tracker to discuss.

I _think_ https://github.com/fastify/fastify/pull/2166 is supposed to fix that.

**Comment by Eomm** [comment]
I'll try to explain what is happening

- v2:

this `vacation: 'sharedAddress#'`, at runtime, become `vacation: { a json }`.
The string is replaced with the right object.
NB: v2 supports `$ref` too, and it would **not** work as per v3.

- in v3:

this code `{ $ref: 'sharedAddress#' }` remains this `{ $ref: 'sharedAddress#' }`
`fastify-swagger` doesn't manage the `$ref` nor inject the shared schema to `swagger-ui`

"by design" we remove the non standard "replace schema feature", here all the details https://github.com/fastify/fastify/pull/2023

----

solutions for @adrai :

- slow: update fastify-swagger (fastify-oas should have the same issue). I know it and I have in plan to do so since I want/need to improve a lot this plugin
- fast: use `fastify.getSchema('sharedAddress')` instead of `$ref` (read [here how](https://dev.to/eomm/validation-and-serialization-in-fastify-v3-2e8l) )

**Comment by adrai** [comment]
@Eomm thank you for the explanation.
I would prefer "slow" (I have no time pressure to update to fastify v3 currently, just did a first test)

**Comment by SkeLLLa** [comment]
Some ideas.

I had an idea to make behaviour that "addModels" do default, so if you use addSchema, then it will appear in OpenAPI spec. But when schema is added it's hard to determine in which place it should be stored (`#/components/schemas/` or `#/components/parameters/`). So @Eomm PR give me an idea how to do that.

Instead of adding schemas when they are registered, we can add schemas when we find `$ref` and add it to proper components section https://swagger.io/docs/specification/components/.

1. Scan route schema for `$ref`.
2. Obtain schema from some "cache" or from "getSchemasRecursive".*
3. General rule - get schema and put it into proper section with given name or maybe with instance prefix + name. In case if it's already stored there - just skip (or maybe compare it and error if they are different).
3.1. `$ref` in "root" level means that we should reference the whole item (e.g. body, response, ) 
```
{
  body: { $ref: 'my-body' } // it will contain description and so on
}
```
3.1.1 If parsing body, then store in `#/components/requestBodies/`.
3.1.2 If parsing responses, then store in `#/components/responses/`.
3.1.3 If parsing path, query, header, cookie, then store in `#/components/parameters/`.
3.1.4 If parsing examples, then store in `#/components/examples/`.

3.2. `$ref` in non-root level (e.g. when it describes data models)

```
{
  body: { 
    type: object,
    description: 'my body'
    properties: { 
       foo: {
          $oneOf: [{ $ref: 'cat' }, { $ref: 'dog' }],
       }
    }
  }
}
```
3.2.1 Data models should be added to `#/components/schemas/`.

Also there's other option that could replace [2]. When each instance is registered, this plugin should add a route that will expose schemas that it adds. And use "Remote Reference". If we have some "service" that's registered via `fastify.register` and it has prefix `/foo`, then we need a route
/foo/__schemas__.json and put components there. And in main spec we could use reference like
```
{
  $ref: '/foo/__schemas__.json#/components/schema/Dog'
}
```

**Comment by intpp** [comment]
any updates?

**Comment by BnayaZil** [comment]
@SkeLLLa after I update `fastify-swagger` to version `3.1.2` and modify the `fastify: '>=2.0.0 <2.12.0 || >=2.12.1 <4.0.0'` I manage to run my app locally, maybe we can push this change in?

**Comment by SkeLLLa** [comment]
In order to get this work with fastify v3, you'll need to setup fastify-oas@next.

**Comment by fyang1024** [comment]
Hi @SkeLLLa This is probably a noob question. What do you mean `setup fastify-oas@next`? I am using `fastify 3.2.0`, and I get a blank page when accessing the open api doc generated by `fastify-oas 3.0.1`

**Comment by SkeLLLa** [comment]
@fyang1024 it looks like swagger-ui + browser bug. https://github.com/swagger-api/swagger-ui/issues/3370. You can disable CSP in your browser.

**Comment by fyang1024** [comment]
> @fyang1024 it looks like swagger-ui + browser bug. [swagger-api/swagger-ui#3370](https://github.com/swagger-api/swagger-ui/issues/3370). You can disable CSP in your browser.

Indeed, I disabled csp and it shows the doc now, thank you.  But that does not really solve the issue. It seems Swagger UI team does not want to address this issue. I see Java provides such a solution https://github.com/swagger-api/swagger-ui/issues/3370#issuecomment-627932612. I wonder if fastify-oas can do anything similar? I think it's a HTML header right?

I am still curious what is `fastify-oas@next`?

**Comment by jmilldotdev** [comment]
Bump, still a problem for me.
